### DISCUSSION TITLE
support `in` and `!in` for type checks

### DISCUSSION THREAD

**Discussion by SleepyRoy** [vlang/v#11805] [Ideas]
（updated）

Let's write
```V
        if foo in [TypeA, TypeB, TypeC, TypeD]
```
or
```V
        if foo in TypeA | TypeB | TypeC | TypeD
```
or
```V
        if foo is TypeA | TypeB | TypeC | TypeD
```
instead of
```V
        if foo is TypeA || foo is TypeB || foo is TypeC || foo is TypeD
```
Same to `!in`.

**Comment by tomByrer** [comment]
Curious; is there 'prior art'?  What other languages use your suggested syntax?

**Comment by SleepyRoy** [comment]
@tomByrer I'm not sure. But I'm also ok with `is !in`.

**Comment by medvednikov** [comment]
it's `if foo in [TypeA, TypeB, TypeC, TypeD]` in V

**Comment by JalonSolov** [comment]
Don't need the extra `is`.  Just `in` or `!in`.

**Comment by SleepyRoy** [comment]
oh great, I got it. Since V variable use snake_case while type use PascalCase. Just `in` won't cause troubles.

**Comment by ntrel** [comment]
> it's if foo in [TypeA, TypeB, TypeC, TypeD] in V

`[...]` is array literal syntax, it's elements should be values not types. It would be more consistent to use a sum type literal `Foo in TypeA | TypeB | TypeC | TypeD`. It's also simpler for the parser.

**Comment by spaceface777** [comment]
imho, `Foo is TypeA | TypeB | TypeC` is even better, since it is consistent with the sumtype declaration syntax.

<details>
<summary>although...</summary>
it may not be a good idea if we decide that anonymous sumtypes will not be supported, since this may actually look like one.
</details>

**Comment by tomByrer** [comment]
> It would be more consistent to use a sum type literal `Foo in TypeA | TypeB | TypeC | TypeD`. It's also simpler for the parser.

Seems also closer to [TypeScript's Union](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html) syntax, perhaps IMHO slightly better.

**Comment by ntrel** [comment]
> Foo is TypeA | TypeB | TypeC is even better

Foo may *implicitly convert to* that sum type, but `Foo` is not literally that type. So far we have `sum is Type` - a (runtime) test for *exact type equality*, not type conversion. We need `T is Type2` to test type equality (at compile-time). `T in Type | Type2` needs different syntax because it is doing multiple tests.
